internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+	mList     *memberlist.Memberlist
+	mlJoinCh  chan struct{}
+
+	mlMux         sync.Mutex // Mutex for mlSpeakersIPs
+	mlSpeakersIPs []string   // Speaker IPs
+}
+
+// New creates a new SpeakerList
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:         logger,
+		stopCh:    stopCh,
+		namespace: namespace,
+		labels:    labels,
+	}
+
+	if namespace == "" || labels == "" || bindAddr == "" {
+		logger.Log("op", "startup", "msg", "Not starting fast dead node detection (MemberList), need ml-bindaddr / ml-labels / ml-namespace config")
+		return &sl, nil
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
+	mconfig.Name = nodeName
+	mconfig.BindAddr = bindAddr
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret == "" {
+		sl.l.Log("op", "startup", "warning", "no ml-secret-key set, MemberList traffic will not be encrypted")
+	} else {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+
+	// This channel is used by Rejoin(), that is used as a callback on node
+	// changes for k8s watch. It is assumed that operations there don't
+	// block too much time (is single thread), so make it 'big' to reduce
+	// the chances of blocking.
+	// K8s max node size is 5k, and this call back on node rejoin can be
+	// called 2-3 times (node change to Not Ready, then Ready again and
+	// sometimes another change is detected too), so we make it 5k * 3
+	sl.mlJoinCh = make(chan struct{}, 5000*3)
+
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
+	sl.mlEventCh = make(chan memberlist.NodeEvent, 1024)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.mlEventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
```

Comment:
Can't we leverage built-in functionality in the k8s library to avoid looping and waiting? I feel like we are re-implementing things which are provided by client-go.

We should probably look at using an informer (inside the `k8s` package) like we already do for watching other k8s objects.

Reply:
Well, that is what #595 does, uses a service, and the ideas we discussed offline.

I'm not sure there is another way rather than creating a service for the speakers (which is not clear is something we want, as it is only needed for layer 2 and now the user needs to create a service) and then monitoring it. I'm short of time, so if you have the time to check if there are other alternatives, it would be great :)

An alternative in the path just mentioned (monitor a service), is to create a service an monitor it from the speakerlist package. Basically, we instantiate a `internal/k8s.go` monitoring. But we either modify it to accept a selector (currently monitors all services) or do things like get notified when a service changed and if it's not the metalLB speakers, ignore. As I mentioned offline, I think these ideas have merrit, but I really doubt about doing those changes in a patch release.

Basically it means all metalLB users need to create a service, and we haven't thought the design through to know if this is something that will stay or not. I'm really hesistant about it.

We added memberlist without any kind of re-join mechanism and now we need to fix it. I think to be safe, I lean towards fixing it with the primitives we already have. Unless we find a clean and simple way without exposing more things to the user (like having a service) that we might want to delete afterwards.

What do you think?

Reply:
I'm not sure if the client version used by metallb have it. Can we use NewFilteredListWatchFromClient with a list option that has the labelSelector?
https://godoc.org/k8s.io/client-go/tools/cache#NewFilteredListWatchFromClient
https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions

Reply:
If I remember correctly you don't have the watch permission

Reply:
@liuyuan10 cool, thanks! @johananl can you check those when you have some time? :)

Reply:
> I'm not sure if the client version used by metallb have it. Can we use NewFilteredListWatchFromClient with a list option that has the labelSelector?
> https://godoc.org/k8s.io/client-go/tools/cache#NewFilteredListWatchFromClient
> https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions

Great pointers @liuyuan10, thanks!

Looks like we can do the following:

```go
om := func(o *metav1.ListOptions) {
	o.LabelSelector = "foo=bar"
}
watcher := cache.NewFilteredListWatchFromClient(c.client.CoreV1().RESTClient(), "pods", v1.NamespaceAll, om)
```

I'm looking into it.

Reply:
We discussed offline with @johananl and we agreed that adding a watch mechanism without a proper design (so it might change) doesn't seem nice for a patch release. It is too big of a change to be safe, and without a design it seems really unsafe for a patch release.

For 0.9.4 we plan to just poll very gently, as this patch is doing.